pymatviz/_preprocess_data.py

```diff
@@ -0,0 +1,226 @@
+
+    # Get replacement value for missing and infinity
+    values = df_in[col].explode()
+    numeric_values = pd.to_numeric(values, errors="coerce")
+
+    replacement_nan = 0 if missing_strategy == "zero" else numeric_values.mean()
+    replacement_inf_pos = numeric_values[numeric_values != np.inf].max()
+    replacement_inf_neg = numeric_values[numeric_values != -np.inf].min()
+
+    replacements = {
+        np.nan: replacement_nan,
+        np.inf: replacement_inf_pos,
+        -np.inf: replacement_inf_neg,
+    }
+
+    df_in[col] = df_in[col].apply(
+        lambda val: replacements.get(val, val)
+        if isinstance(val, float)
+        else np.array([replacements.get(v, v) for v in val])
+    )
+
+    return df_in
+
+
+def preprocess_ptable_data(
+    data: SupportedDataType,
+    missing_strategy: Literal["zero", "mean"] = "mean",
+) -> pd.DataFrame:
+    """Preprocess input data for ptable plotters, including:
+        - Convert all data types to pd.DataFrame.
+        - Replace missing values (NaN) by selected strategy.
+        - Replace infinities with vmax(∞) or vmin(-∞).
+        - Write vmin/mean/vmax as metadata into the DataFrame.
+
+    Args:
+        data (dict[str, float | Sequence[float]] | pd.DataFrame | pd.Series):
+            Input data to preprocess.
+        missing_strategy: missing value replacement strategy.
+
+    Returns:
+        pd.DataFrame: The preprocessed DataFrame with element names
+            as index and values as columns.
+
+    Example:
+        >>> data_dict: dict = {
+            "H": 1.0,
+            "He": [2.0, 4.0],
+            "Li": [[6.0, 8.0], [10.0, 12.0]],
+        }
+
+        OR
+        >>> data_df: pd.DataFrame = pd.DataFrame(
+            data_dict.items(),
+            columns=["element", "heat_val"]
+            ).set_index("element")
+
+        OR
+        >>> data_series: pd.Series = pd.Series(data_dict)
+
+        >>> preprocess_data(data_dict / df / series)
+
+        >>> data_df
+            Element   Value
+            H         [1.0, ]
+            He        [2.0, 4.0]
+            Li        [[6.0, 8.0], [10.0, 11.0]]
+
+        Metadata (data_df.attrs):
+            vmin: 1.0
+            mean: 6.0
+            vmax: 11.0
+    """
+    if isinstance(data, pd.DataFrame):
+        if Key.element == data.index.name and Key.heat_val in data:
+            data_df = data
+        else:
+            warnings.warn(
+                "DataFrame has unexpected format, trying to fix.", stacklevel=2
+            )
+            try:
+                data_df = pd.melt(
+                    data.reset_index(),
+                    id_vars=data.index.names,
+                    var_name=Key.element,
+                    value_name=Key.heat_val,
+                ).set_index(Key.element)
+            except KeyError as exc:
+                raise KeyError(
+                    f"Cannot normalize the given DataFrame:\n{data!s}"
```

Comment:
I'm a bit concerned with the wording "normalize" here, I understand in this context it means the same as "re-format the DataFrame", but I assume it could be confused with the other meaning "to rescale data to a certain range", especially in the context of "data preprocess"

Reply:
good point. how about "normalize" -> "handle"? maybe we should also add a sentence what needs to change for the dataframe to become valid

Reply:
> good point. how about "normalize" -> "handle"? 

I would personally prefer my original wording of "fix" :), but either way both should be pretty clear :)

> maybe we should also add a sentence what needs to change for the dataframe to become valid

Thanks for adding this!